node/pkg/db/redis.go

```diff
@@ -23,106 +24,151 @@ type RedisConnectionInfo struct {
 
 var (
 	rdbMutex sync.Mutex
-	rdb      *redis.Conn
+	rdb      *redis.Client
 )
 
-func GetRedisConn(ctx context.Context) (*redis.Conn, error) {
-	conn, err := getRedisConn(ctx)
-	if err != nil {
-		return nil, err
-	}
-	return conn, err
-}
-
-func getRedisConn(ctx context.Context) (*redis.Conn, error) {
+func GetRedisConn(ctx context.Context) (*redis.Client, error) {
 	rdbMutex.Lock()
 	defer rdbMutex.Unlock()
```

Comment:
I [read](https://github.com/redis/go-redis/pull/167/files#diff-190fc9ceda3bbdbe723d493b8b99e2b9a0100c7d635bfe9b04246280bef6f67eR181) that redis client can handle concurrent access. So I wanted to suggest refactoring `GetRedisConn` function to make it partially mutex:
```go
func GetRedisConn(ctx context.Context) (*redis.Client, error) {
    if rdb != nil {
        return rdb, nil
    }

    rdbMutex.Lock()
    defer rdbMutex.Unlock()

    // Check the condition again inside the lock to ensure another routine hasn't already initialized the connection
    if rdb == nil {
        err := reconnectRedis(ctx)
        return rdb, err
    }

    return rdb, nil
}
```
this way when the connection is alive concurrent accesses wont be blocked and might improve performance

Reply:
great suggestion, will do

Reply:
reading the value also requires mutex lock so I'll keep it as it is

Reply:
made some updates to use RWMutex instead based on your suggestion